consensus/src/threshold_simplex/types.rs

```diff
@@ -64,45 +54,284 @@ pub trait Attributable {
+        self.signer.encode_size() + self.signature.encode_size()
+    }
+}
+
+impl<S: SigningScheme> Read for Vote<S> {
+    type Cfg = ();
+
+    fn read_cfg(reader: &mut impl Buf, _: &()) -> Result<Self, Error> {
+        let signer = u32::read(reader)?;
+        let signature = S::Signature::read(reader)?;
+
+        Ok(Self { signer, signature })
+    }
+}
+
+/// Result of verifying a batch of votes.
+pub struct VoteVerification<S: SigningScheme> {
+    /// Contains the votes accepted by the scheme.
+    pub verified: Vec<Vote<S>>,
+    /// Identifies the participant indices rejected during batch verification.
+    pub invalid_signers: Vec<u32>,
+}
+
+impl<S: SigningScheme> VoteVerification<S> {
+    /// Creates a new `VoteVerification` result.
+    pub fn new(verified: Vec<Vote<S>>, invalid_signers: Vec<u32>) -> Self {
+        Self {
+            verified,
+            invalid_signers,
+        }
+    }
 }
 
-// Constants for domain separation in signature verification
-// These are used to prevent cross-protocol attacks and message-type confusion
-pub const SEED_SUFFIX: &[u8] = b"_SEED";
-pub const NOTARIZE_SUFFIX: &[u8] = b"_NOTARIZE";
-pub const NULLIFY_SUFFIX: &[u8] = b"_NULLIFY";
-pub const FINALIZE_SUFFIX: &[u8] = b"_FINALIZE";
+/// Cryptographic surface required by `threshold_simplex`.
+///
+/// A `SigningScheme` produces validator votes, validates them (individually or in
+/// batches), assembles quorum certificates, checks recovered certificates and, when
+/// available, derives randomness for leader rotation. Implementations may override the
+/// provided defaults to take advantage of scheme-specific batching strategies.
+pub trait SigningScheme: Clone + Debug + Send + Sync + 'static {
+    type Signature: Clone
+        + Debug
+        + PartialEq
+        + Eq
+        + Hash
+        + Send
+        + Sync
+        + EncodeSize
+        + Write
+        + Read<Cfg = ()>;
+
+    type Certificate: Clone
+        + Debug
+        + PartialEq
+        + Eq
+        + Hash
+        + Send
+        + Sync
+        + EncodeSize
+        + Write
+        + Read<Cfg = Self::CertificateCfg>;
+
+    type Randomness: Clone + EncodeSize + Write + Send;
+
+    type CertificateCfg: Clone + Send + Sync;
+
+    /// Converts the scheme into a pure verifier.
+    ///
+    /// The returned instance should return `false` from `can_sign()`.
+    fn into_verifier(self) -> Self;
+
+    /// Returns `true` if this instance holds the secrets required to author votes.
+    fn can_sign(&self) -> bool;
+
+    /// Signs a vote for the given context using the supplied namespace for domain separation.
+    fn sign_vote<D: Digest>(&self, namespace: &[u8], context: VoteContext<'_, D>) -> Vote<Self>;
+
+    /// Verifies a single vote against the participant material managed by the scheme.
+    fn verify_vote<D: Digest>(
+        &self,
+        namespace: &[u8],
+        context: VoteContext<'_, D>,
+        vote: &Vote<Self>,
+    ) -> bool;
```

Comment:
Thinking out loud... consider two different traits. `trait SchemeVerifier: ... { ... }` and `trait SchemeSigner: SchemeVerifier { ... }`.

`into_verifier` can be a function only on `SchemeSigner` and will return a type of `SchemeVerifier`.